File: Another Contest 3 Problem 1 - Diverse Arrays.cpp
// need fast input, direct translation of my python code that gets TLE

#include <iostream>
#include <unordered_map>
using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(NULL);

    int n, k;
    cin >> n >> k;
    int nums[n];
    for (int i = 0; i < n; i++) {
        cin >> nums[i];
    }

    long long total = (long long)n * (n + 1) / 2;  // total number of sub arrays

    unordered_map<int, int> freq;
    int left = 0;  // sliding window
    for (int right = 0; right < n; right++) {
        freq[nums[right]]++;

        while (freq.size() >= k) {
            int left_num = nums[left];
            freq[left_num]--;
            if (freq[left_num] == 0) {
                freq.erase(left_num);
            }
            left++;
        }

        total -= (right - left + 1);
    }

    cout << total << endl;
}

File: Another Contest 3 Problem 1 - Diverse Arrays.py
# TLE, check c++ solution

import sys
from collections import defaultdict

input = sys.stdin.readline  # need this, up to 1 mil lines of input

n, k = map(int, input().split())
nums = [int(input()) for _ in range(n)]

total = (n * (n + 1)) // 2  # total number of sub arrays

freq = defaultdict(int)
left = 0  # sliding window
for right in range(n):
    freq[nums[right]] += 1

    while len(freq) >= k:
        left_num = nums[left]
        freq[left_num] -= 1
        if freq[left_num] == 0:
            del freq[left_num]
        left += 1

    total -= (right - left + 1)

print(total)
